README.md
# talos-pxe

Talos PXE is a project aimed at boostrapping an initial Talos cluster from a single bootable USB stick. This stick plugs into an avaiable computer and will provide a selection menu for any PXE bootable machine connected to the same network. The software is able to adapt to an existing DHCP server via proxyDHCP or provide a DHCP server of its own.

![iPXE Talos menu screenshot](screenshot.png)

## Requirements

- Docker
- Golang version 1.16.7
- LinuxKit

## Building

You need to download the Talos kernel and initramfs before proceeding. Files can be found on [talos release page](https://github.com/talos-systems/talos/releases) These should be placed in `assets`:

```
assets/initramfs-amd64.xz
assets/vmlinuz-amd64
```

Generate the configuration via:

```
talosctl gen config -o assets talos-k8s-metal-tutorial https://controlplane.talos:8443
sed 's/type: controlplane/type: init/' assets/controlplane.yaml > assets/init.yaml
```

First step is building the pxe network container via:

```
docker build -t talos-pxe --target talos-pxe .
```

then that container can be converted into a bootable VM via LinuxKit:

```
linuxkit build -docker -format iso-bios linux.yml

(if you get a VFS error booting this image, you may need to try other formats like `raw-bios` or `raw-efi`)
```

## Unittests

Unittests are run in a docker container, that is build before the tests, so they can be run on Linux or Mac. (I have not tried 
Windows). Coverage files can be found after run in out/coverage.html.

```bash
make unittest
```

One test can be run with following make target:

```bash
make unittest-one TEST_PATTERN=<test name>
```

## Local verification setup

Disclaimer: This setup will probably not work for you, but it can remind me in a few months how to set this thing up.
Used virtualization `libvirt + KVM`.

### Prepare binary and served files

1. Follow steps from build section, about assets and talos configs. 
2. Set required capabilities on the talos-pxe binary:
    ```bash
    sudo setcap cap_net_raw,cap_net_admin,cap_net_bind_service+eip ./talos-pxe
    ```

### Setup for verifying DHCP no proxy mode

1. Stop DHCP service on the virbr0 interface, find ip range used on that interface, it is 192.168.122.1/24 in my case, then find DHCP server on it:
    ```bash
    sudo netstat -lutnp
    
    Active Internet connections (only servers)
    Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
    tcp        0      0 192.168.122.1:53        0.0.0.0:*               LISTEN      1216/dnsmasq         
    ...
    udp        0      0 192.168.122.1:53        0.0.0.0:*                           1216/dnsmasq        
    ... 
    udp        0      0 0.0.0.0:67              0.0.0.0:*                           1216/dnsmasq

    sudo kill -9 <pid for dnsmasq>
    ```
2. You might need to allow ports needed by talos-pxe to be open, I did not have to on Ubuntu 20.04.
3. Run talos-pxe:
   ```bash /talos-pxe --if virbr0 --addr 192.168.122.1/24```
4. In virt-manager create a new vm with PXE boot and watch it working.

### Setup for verifying DHCP proxy mode
1. If you were runing the no proxy mode first then get the dnsmasq back up.
2. We need to set up an interface that is controlled by talos-pxe and connected to the virbr0 interface tha other DHCP server is controlling, (dnsmasq in my case).
to get there we are doing some magic with a veth pair: 
    ```bash
    sudo ip link add eth0 type veth peer name eth1
    sudo ip link set dev eth1 master virbr0
    sudo ip link set eth1 up
    ```
3. Again you might need to set some firewall rules for the ports used by talos-pxe, I did not have to on Ubuntu 20.04, On RHEL/Fedora do:
   ```bash
   sudo firewall-cmd --add-port=4011/tcp --zone=libvirt
   sudo firewall-cmd --add-interface=eth0 --zone=libvirt
   sudo firewall-cmd --add-interface=eth1 --zone=libvirt
   ```
4. Run talos pxe with that interface:
```bash ./talos-pxe --if eth0 ```

 


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.1] 2022-01-10

- Added unit tests 

## [0.1.0] 2021-12-30

- It works I think 


dhcp_test.go
package main

import (
	"net"
	"testing"

	"github.com/stretchr/testify/require"
)

func Test_getAvailableRange(t *testing.T) {
	tests := map[string]struct {
		cidr  string
		want  string
		want1 string
	}{
		"default case": {
			"192.168.123.1/24", "192.168.123.2", "192.168.123.254",
		},
		"network 8": {
			"10.0.0.1/8", "10.0.0.2", "10.255.255.254",
		},
	}
	for name, tt := range tests {
		t.Run(name, func(t *testing.T) {
			netIp, ipNet, err := net.ParseCIDR(tt.cidr)
			require.Nil(t, err)

			got, got1 := getAvailableRange(*ipNet, netIp)
			require.Equal(t, tt.want, got.String())
			require.Equal(t, tt.want1, got1.String())
		})
	}
}


tftp_test.go
package main

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"testing"

	"github.com/pin/tftp"
	"github.com/stretchr/testify/require"
)

const (
	expectedInPxeMenu = `
:init
chain http://127.0.0.1:8080/ipxe?uuid=${uuid}&ip=${ip}&mac=${mac:hexhyp}&domain=${domain}&hostname=${hostname}&serial=${serial}&type=init

:controlplane
chain http://127.0.0.1:8080/ipxe?uuid=${uuid}&ip=${ip}&mac=${mac:hexhyp}&domain=${domain}&hostname=${hostname}&serial=${serial}&type=controlplane

:worker
chain http://127.0.0.1:8080/ipxe?uuid=${uuid}&ip=${ip}&mac=${mac:hexhyp}&domain=${domain}&hostname=${hostname}&serial=${serial}&type=worker`
)

func TestTFTPHook(t *testing.T) {
	/*
		Simple test that TFTPHook does log correct message
	*/
	capture, cleanup := NewLogCapture()
	defer cleanup()

	tHook := &TFTPHook{}

	stats := tftp.TransferStats{
		Filename: "foo", RemoteAddr: net.IPv4(1, 2, 3, 4),
	}
	tHook.OnSuccess(stats)
	tHook.OnFailure(stats, errors.New("boom"))
	capture.RequireInLog(t, "Transferred foo to 1.2.3.4")
	capture.RequireInLog(t, "Failure transferring foo to 1.2.3.4: boom")
}

func TestServeTFTP(t *testing.T) {
	/**
	Test behaviour of the tftp server by running it and querying the tftp client
	*/
	s, cleanup := talosPxeServerForTest(t, true)
	defer cleanup()

	require.NotNil(t, s)

	c, err := tftp.NewClient("127.0.0.1:69")
	require.Nil(t, err)

	// When iPXE is in classInfo then we expect to get the pxe menu
	wt, err := c.Receive("98:01:a7:99:ac:6b/whatever/iPXE", "octet")
	require.Nil(t, err)

	var buff bytes.Buffer
	_, err = wt.WriteTo(&buff)
	require.Nil(t, err)
	require.Contains(t, buff.String(), expectedInPxeMenu)

	// When iPXE is not classInfo then we will expect to get the pxe file
	for _, option := range []string{
		"98:01:a7:99:ac:6b/PXEClient:Arch:00000:UNDI:002001/what",
		"98:01:a7:99:ac:6b/PXEClient:Arch:00007:UNDI:003001/what"} {
		wt, err := c.Receive(option, "octet")
		require.Nil(t, err)

		var buff bytes.Buffer
		_, err = wt.WriteTo(&buff)
		require.Nil(t, err)
		require.Equal(t, "my fake fakeIpxe", buff.String())
	}

	// Test error conditions, unknown class, wrong mac address and wrong path
	for _, option := range []string{
		"not a mac address/PXEClient:Arch:00000:UNDI:002001/what",
		"98:01:a7:99:ac:6b/unknown class/what",
		"98:01:a7:99:ac:6b/not enough of path elements"} {
		_, err := c.Receive(option, "octet")
		require.NotNil(t, err)

		fmt.Printf("Error %v", err)
	}
}


server_test.go
package main

import (
	"fmt"
	"net"
	"net/http"
	"net/http/httptest"
	"net/url"
	"testing"

	"github.com/stretchr/testify/require"
)

type myTestIpxeWrapperMenuHandler struct {
	paths []string
	t     *testing.T
}

func (h *myTestIpxeWrapperMenuHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
	h.paths = append(h.paths, req.URL.Path)

	// for requests for ipxe path we expect that writer is ResponseRecorder
	if req.URL.Path == "ipxe" || req.URL.Path == "/ipxe" {
		_, ok := w.(*httptest.ResponseRecorder)
		require.True(h.t, ok)
		if req.URL.Path == "ipxe" {
			// return 404 so we can test that menu is served from ipxeWrapperMenuHandler
			w.WriteHeader(404)
			return
		}
	}
}

func Test_ipxeWrapperMenuHandler(t *testing.T) {
	s, cleanup := talosPxeServerForTest(t, false)
	defer cleanup()
	primaryHandler := myTestIpxeWrapperMenuHandler{t: t}
	handler := s.ipxeWrapperMenuHandler(&primaryHandler)

	t.Run("check requests to not ipxe are handled", func(t *testing.T) {
		writer := myWriter{}
		req := http.Request{URL: &url.URL{Path: "foo"}}
		handler.ServeHTTP(&writer, &req)
		require.Len(t, primaryHandler.paths, 1)
		require.Equal(t, "foo", primaryHandler.paths[0])
	})

	t.Run("check that recorder is used for ipxe requests and on error we return the menu", func(t *testing.T) {
		capture, cleanupLogger := NewLogCapture()
		defer cleanupLogger()
		writer := myWriter{}
		req := http.Request{URL: &url.URL{Path: "ipxe"}}
		handler.ServeHTTP(&writer, &req)
		capture.RequireInLog(t, "Serving menu")
		require.Contains(t, string(writer.Data), expectedInPxeMenu)
	})

	t.Run("Error parsing form", func(t *testing.T) {
		capture, cleanupLogger := NewLogCapture()
		defer cleanupLogger()
		writer := myWriter{}
		// /ipxe does not cause error in my handler
		req := http.Request{URL: &url.URL{Path: "/ipxe", RawQuery: "%^"}}
		handler.ServeHTTP(&writer, &req)
		capture.RequireInLog(t, "Error ParseForm: invalid URL escape")
	})

	t.Run("Call to init is registered with dns server", func(t *testing.T) {
		capture, cleanupLogger := NewLogCapture()
		defer cleanupLogger()
		writer := myWriter{}
		// /ipxe does not cause error in my handler
		req := http.Request{URL: &url.URL{Path: "/ipxe", RawQuery: "type=init&ip=1.2.3.4"}}
		handler.ServeHTTP(&writer, &req)
		capture.RequireInLog(t, "Selecting init for 1.2.3.4")
		require.Equal(t, "1.2.3.4", s.DNSRecordsv4[s.Controlplane][0].String())
	})

	t.Run("Call to controlplane is registered with dns server", func(t *testing.T) {
		capture, cleanupLogger := NewLogCapture()
		defer cleanupLogger()
		writer := myWriter{}
		// /ipxe does not cause error in my handler
		req := http.Request{URL: &url.URL{Path: "/ipxe", RawQuery: "type=controlplane&ip=1.2.3.5"}}
		handler.ServeHTTP(&writer, &req)
		capture.RequireInLog(t, "Selecting controlplane for 1.2.3.5")
		require.Equal(t, "1.2.3.5", s.DNSRecordsv4[s.Controlplane][1].String())
	})

	t.Run("Call to worker is NOT registered with dns server", func(t *testing.T) {
		capture, cleanupLogger := NewLogCapture()
		defer cleanupLogger()
		writer := myWriter{}
		// /ipxe does not cause error in my handler
		req := http.Request{URL: &url.URL{Path: "/ipxe", RawQuery: "type=worker&ip=1.2.3.7"}}
		handler.ServeHTTP(&writer, &req)
		capture.RequireInLog(t, "Selecting worker for 1.2.3.7")
		require.Len(t, s.DNSRecordsv4[s.Controlplane], 2)
	})
}

func Test_getInterface(t *testing.T) {
	privAddr, err := getPrivateAddress()
	require.Nil(t, err)
	require.NotEqualf(t, "", privAddr, "Private address should not be empty")
	log.Infof("Private address is %s", privAddr)

	interf, netIpMas, err := getInterface(privAddr)
	require.Nil(t, err)
	require.NotNil(t, interf)
	require.NotNil(t, netIpMas)
	log.Infof("Interface: %v, netMask: %v", interf, netIpMas)

	interf, netIpMas, err = getInterface(net.IP{})
	require.NotNil(t, err)
	require.Nil(t, interf)
	require.Nil(t, netIpMas)
}

func TestLogInfo(t *testing.T) {
	s, cleanup := talosPxeServerForTest(t, false)
	defer cleanup()
	capture, cleanupLogger := NewLogCapture()
	defer cleanupLogger()
	msg := "This is stupid but coverage looks good so meh"
	s.logInfo(msg)
	capture.RequireInLog(t, msg)
}

func talosPxeServerForTest(t *testing.T, startTFTP bool) (*Server, func()) {
	tmpDir := NewTempDir(t, "talosPxeServerForTest")
	current := ipxeFileName
	ipxeFileName = "fakeIpxe"
	_ = tmpDir.Write(ipxeFileName, "my fake fakeIpxe")

	s, err := NewServer(net.IPv4(127, 0, 0, 1), tmpDir.path, "lo", defaultControlplane)
	require.Nil(t, err)

	if startTFTP {
		// TFTP server has to be started because if it is not started then calling Shutdown() is throwing panic
		// as the tftp is closing connection that does not exist
		tftpListener, err := net.ListenPacket("udp", fmt.Sprintf("%s:%d", s.IP, s.TFTPPort))
		require.Nil(t, err)
		go func() {
			err := s.serveTFTP(tftpListener)
			require.Nil(t, err)
		}()
	}

	cleanup := func() {
		ipxeFileName = current
		tmpDir.Cleanup()
		s.Shutdown()
	}
	return s, cleanup
}

// we need a fake writer because one of the tested behaviours is about httptest.ResponseRecorder
// so we can not use it
type myWriter struct {
	Code int
	Data []byte
}

func (m *myWriter) Header() http.Header {
	return http.Header{}
}

func (m *myWriter) Write(bytes []byte) (int, error) {
	m.Data = append(m.Data, bytes...)
	return len(bytes), nil
}

func (m *myWriter) WriteHeader(statusCode int) {
	m.Code = statusCode
}


dns.go
package main

import (
	"context"
	"fmt"
	"net"

	"github.com/coredns/coredns/core/dnsserver"
	"github.com/coredns/coredns/plugin"
	"github.com/coredns/coredns/plugin/forward"
	"github.com/coredns/coredns/plugin/pkg/dnsutil"
	"github.com/coredns/coredns/plugin/pkg/fall"
	"github.com/coredns/coredns/request"
	"github.com/miekg/dns"
)

const (
	DNSTTL = 60
)

type ServiceLookupPlugin struct {
	Next   plugin.Handler
	Fall   fall.F
	Server *Server
	Zones  []string
}

func (s ServiceLookupPlugin) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
	log.Debugf("DNS request: %v", r)
	state := request.Request{W: w, Req: r}
	qname := state.Name()

	answers := []dns.RR{}

	zone := plugin.Zones(s.Zones).Matches(qname)
	if zone == "" {
		if state.QType() != dns.TypePTR {
			return plugin.NextOrFailure(s.Name(), s.Next, ctx, w, r)
		}
	}

	switch state.QType() {
	case dns.TypePTR:
		names := s.GetREntry(dnsutil.ExtractAddressFromReverse(qname))
		if len(names) == 0 {
			return plugin.NextOrFailure(s.Name(), s.Next, ctx, w, r)
		}
		answers = ptr(qname, DNSTTL, names)
	case dns.TypeA:
		ips := s.GetHostV4(qname)
		answers = a(qname, DNSTTL, ips)
	case dns.TypeAAAA:
		ips := s.GetHostV6(qname)
		answers = aaaa(qname, DNSTTL, ips)
	}

	// Only on NXDOMAIN we will fallthrough.
	if len(answers) == 0 && !s.otherRecordsExist(qname) {
		if s.Fall.Through(qname) {
			return plugin.NextOrFailure(s.Name(), s.Next, ctx, w, r)
		}

		return dns.RcodeServerFailure, nil
	}

	m := new(dns.Msg)
	m.SetReply(r)
	m.Authoritative = true
	m.Answer = answers

	w.WriteMsg(m)

	return dns.RcodeSuccess, nil
}

func (s ServiceLookupPlugin) otherRecordsExist(qname string) bool {
	if len(s.GetHostV4(qname)) > 0 {
		return true
	}
	if len(s.GetHostV6(qname)) > 0 {
		return true
	}
	return false
}

func (s ServiceLookupPlugin) GetREntry(ip string) []string {
	s.Server.DNSRWLock.RLock()
	defer s.Server.DNSRWLock.RUnlock()
	return s.Server.DNSRRecords[ip]
}

func (s ServiceLookupPlugin) GetHostV4(name string) []net.IP {
	s.Server.DNSRWLock.RLock()
	defer s.Server.DNSRWLock.RUnlock()
	return s.Server.DNSRecordsv4[name]
}

func (s ServiceLookupPlugin) GetHostV6(name string) []net.IP {
	s.Server.DNSRWLock.RLock()
	defer s.Server.DNSRWLock.RUnlock()
	return s.Server.DNSRecordsv6[name]
}

func (s ServiceLookupPlugin) Name() string {
	return "servicelookupplugin"
}

// a takes a slice of net.IPs and returns a slice of A RRs.
func a(zone string, ttl uint32, ips []net.IP) []dns.RR {
	answers := make([]dns.RR, len(ips))
	for i, ip := range ips {
		r := new(dns.A)
		r.Hdr = dns.RR_Header{Name: zone, Rrtype: dns.TypeA, Class: dns.ClassINET, Ttl: ttl}
		r.A = ip
		answers[i] = r
	}
	return answers
}

// aaaa takes a slice of net.IPs and returns a slice of AAAA RRs.
func aaaa(zone string, ttl uint32, ips []net.IP) []dns.RR {
	answers := make([]dns.RR, len(ips))
	for i, ip := range ips {
		r := new(dns.AAAA)
		r.Hdr = dns.RR_Header{Name: zone, Rrtype: dns.TypeAAAA, Class: dns.ClassINET, Ttl: ttl}
		r.AAAA = ip
		answers[i] = r
	}
	return answers
}

// ptr takes a slice of host names and filters out the ones that aren't in Origins, if specified, and returns a slice of PTR RRs.
func ptr(zone string, ttl uint32, names []string) []dns.RR {
	answers := make([]dns.RR, len(names))
	for i, n := range names {
		r := new(dns.PTR)
		r.Hdr = dns.RR_Header{Name: zone, Rrtype: dns.TypePTR, Class: dns.ClassINET, Ttl: ttl}
		r.Ptr = dns.Fqdn(n)
		answers[i] = r
	}
	return answers
}

func (s *Server) registerDNSEntry(entry string, ip net.IP) {
	s.DNSRWLock.Lock()
	defer s.DNSRWLock.Unlock()
	records := s.DNSRecordsv4[entry]
	for _, r := range records {
		if r.Equal(ip) {
			return
		}
	}
	records = append(records, ip)
	s.DNSRecordsv4[entry] = records
}

func (s *Server) serveDNS(l net.PacketConn) error {
	return s.serverDNS.ServePacket(l)
}

func (s *Server) prepConfDNS() []*dnsserver.Config {
	zone := "talos."

	zoneConfig := &dnsserver.Config{
		Zone:        zone,
		Transport:   "dns",
		ListenHosts: []string{""},
		Port:        fmt.Sprintf("%d", s.DNSPort),
		Debug:       true,
	}

	zoneConfig.AddPlugin(func(next plugin.Handler) plugin.Handler {
		serviceLookup := ServiceLookupPlugin{
			Server: s,
			Zones:  []string{zone},
		}
		serviceLookup.Next = next

		return serviceLookup
	})

	proxyConfig := &dnsserver.Config{
		Zone:        ".",
		Transport:   "dns",
		ListenHosts: []string{""},
		Port:        fmt.Sprintf("%d", s.DNSPort),
		Debug:       true,
	}

	proxyConfig.AddPlugin(func(next plugin.Handler) plugin.Handler {
		forwardProxy := forward.New()
		for _, forwardDns := range s.ForwardDns {
			forwardProxy.SetProxy(forward.NewProxy(forwardDns, "dns"))
		}
		forwardProxy.Next = next

		return forwardProxy
	})

	return []*dnsserver.Config{zoneConfig, proxyConfig}
}


pxe.go
// Copyright 2016 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"fmt"
	"net"

	"github.com/insomniacslk/dhcp/dhcpv4"
	"golang.org/x/net/ipv4"
)

// TODO: this may actually be the BINL protocol, a
// Microsoft-proprietary fork of PXE that is more universally
// supported in UEFI than PXE itself. Need to comb through the
// TianoCore EDK2 source code to figure out if what this is doing is
// actually BINL, and if so rename everything.

// TODO test it, to unit test it i need examples of binary representation of some pxe requests

func (s *Server) servePXE(conn net.PacketConn) error {
	buf := make([]byte, 1024)
	l := ipv4.NewPacketConn(conn)
	if err := l.SetControlMessage(ipv4.FlagInterface, true); err != nil {
		return fmt.Errorf("Couldn't get interface metadata on PXE port: %s", err)
	}

	for {
		select {
		case <-s.closeServers:
			log.Info("PXE server stopping")
			return nil
		default:
			// continue
		}
		n, msg, addr, err := l.ReadFrom(buf)
		if err != nil {
			return fmt.Errorf("Receiving packet: %s", err)
		}

		log.Infof("Received proxyDHCP PXE request from %s", addr)

		m, err := dhcpv4.FromBytes(buf[:n])
		if err != nil {
			log.Debugf("Packet from %s is not a DHCP packet: %s", addr, err)
			continue
		}

		if m.OpCode != dhcpv4.OpcodeBootRequest || !m.IsOptionRequested(dhcpv4.OptionBootfileName) {
			log.Debugf("Ignoring packet from %s (%s): %s", m.ClientHWAddr, addr, err)
			continue
		}

		resp, err := dhcpv4.NewReplyFromRequest(m,
			dhcpv4.WithOption(dhcpv4.OptMessageType(dhcpv4.MessageTypeAck)),
			dhcpv4.WithOption(dhcpv4.OptBootFileName(fmt.Sprintf("%s/%s/%s", m.ClientHWAddr, m.ClassIdentifier(), m.UserClass()))),
			dhcpv4.WithOption(dhcpv4.OptServerIdentifier(s.IP)),
			dhcpv4.WithOption(dhcpv4.OptGeneric(dhcpv4.OptionClassIdentifier, []byte("PXEClient"))),
		)
		resp.ServerIPAddr = s.IP

		if m.Options[dhcpv4.OptionClientMachineIdentifier.Code()] != nil {
			resp.UpdateOption(dhcpv4.OptGeneric(dhcpv4.OptionClientMachineIdentifier, m.Options[dhcpv4.OptionClientMachineIdentifier.Code()]))
		}

		log.Debug(resp.Summary())
		if _, err := l.WriteTo(resp.ToBytes(), &ipv4.ControlMessage{
			IfIndex: msg.IfIndex,
		}, addr); err != nil {
			log.Errorf("Failed to send PXE response to %s (%s): %s", m.ClientHWAddr, addr, err)
		}
	}
}


dhcp.go
// Copyright 2016 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"encoding/binary"
	"fmt"
	"net"
	"time"

	"github.com/insomniacslk/dhcp/dhcpv4"
	"github.com/insomniacslk/dhcp/dhcpv4/server4"
	"github.com/insomniacslk/dhcp/iana"
)

func (s *Server) handlerDHCP4() server4.Handler {
	leaseTime := 5 * time.Minute

	return func(conn net.PacketConn, peer net.Addr, m *dhcpv4.DHCPv4) {
		log.Debugf("DHCPv4: got %s", m.Summary())

		if m.OpCode != dhcpv4.OpcodeBootRequest {
			log.Infof("Not a boot request")
			return
		}

		resp, err := dhcpv4.NewReplyFromRequest(m,
			dhcpv4.WithOption(dhcpv4.OptServerIdentifier(s.IP)),
		)
		if err != nil {
			log.Error(err)
			return
		}

		efi := false
		ipxe := false

		for _, a := range m.ClientArch() {
			switch {
			case a == iana.EFI_ITANIUM:
				fallthrough
			case a >= iana.EFI_IA32 && a <= iana.EFI_ARM64:
				fallthrough
			case a >= iana.EFI_X86_HTTP && a <= iana.EFI_ARM64_HTTP:
				fallthrough
			case a >= iana.EFI_RISCV32 && a <= iana.EFI_RISCV128_HTTP:
				fallthrough
			case a >= iana.EFI_MIPS32 && a <= iana.EFI_SUNWAY64:
				efi = true
			}
		}

		for _, u := range m.UserClass() {
			switch {
			case u == "iPXE":
				ipxe = true
			}
		}

		if !s.ProxyDHCP {
			s.DHCPLock.Lock()
			defer s.DHCPLock.Unlock()

			record, ok := s.DHCPRecords[m.ClientHWAddr.String()]
			if !ok {
				newIp, err := s.DHCPAllocator.Allocate(net.IPNet{})
				if err != nil {
					log.Error(err)
					return
				}

				record = &DHCPRecord{
					IP:      newIp.IP,
					expires: time.Now().Add(leaseTime),
				}
				s.DHCPRecords[m.ClientHWAddr.String()] = record

			} else {
				if record.expires.Before(time.Now().Add(leaseTime)) {
					record.expires = time.Now().Add(leaseTime).Round(time.Second)
				}
			}

			resp, err = dhcpv4.NewReplyFromRequest(m,
				dhcpv4.WithNetmask(s.Net.Mask),
				dhcpv4.WithYourIP(record.IP),
				dhcpv4.WithGatewayIP(s.GWIP),
				dhcpv4.WithOption(dhcpv4.OptRouter(s.GWIP)),
				dhcpv4.WithOption(dhcpv4.OptIPAddressLeaseTime(leaseTime)),
				dhcpv4.WithOption(dhcpv4.OptServerIdentifier(s.IP)),
			)
			if err != nil {
				log.Error(err)
				return
			}
		} else {
			resp.UpdateOption(dhcpv4.OptGeneric(dhcpv4.OptionClassIdentifier, []byte("PXEClient")))

			if m.Options[dhcpv4.OptionClientMachineIdentifier.Code()] != nil {
				resp.UpdateOption(dhcpv4.OptGeneric(dhcpv4.OptionClientMachineIdentifier, m.Options[dhcpv4.OptionClientMachineIdentifier.Code()]))
			}

			// Some EFI firmwares refuse to boot if PXE Boot Server Discovery Control is set, so
			// only set it if not on EFI
			if !efi {
				pxe := []byte{
					// PXE Boot Server Discovery Control - bypass, just boot from filename.
					6, 1, 8, byte(dhcpv4.OptionEnd),
				}

				resp.UpdateOption(dhcpv4.OptGeneric(dhcpv4.OptionVendorSpecificInformation, pxe))
			}
		}

		resp.Options.Update(dhcpv4.OptDNS(s.IP))
		resp.ServerIPAddr = s.IP

		if m.IsOptionRequested(dhcpv4.OptionBootfileName) {
			log.Infof("received PXE boot request from %s", m.ClientHWAddr)

			log.Infof("sending PXE response to %s", m.ClientHWAddr)

			resp.UpdateOption(dhcpv4.OptTFTPServerName(s.IP.String()))

			if ipxe {
				// In proxyDHCP, iPXE ignores TFTPServerName option if DHCP sent it, so we have to use tftp://
				resp.UpdateOption(dhcpv4.OptBootFileName(fmt.Sprintf("tftp://%s/%s/%s/%s", s.IP, m.ClientHWAddr, m.ClassIdentifier(), m.UserClass())))
			} else {
				// other clients don't understand tftp://, but they will accept TFTPServerName, even in proxyDHCP
				resp.UpdateOption(dhcpv4.OptBootFileName(fmt.Sprintf("%s/%s/%s", m.ClientHWAddr, m.ClassIdentifier(), m.UserClass())))
			}
		}

		//resp.UpdateOption(dhcpv4.OptGeneric(dhcpv4.OptionInterfaceMTU, dhcpv4.Uint16(match.MTU).ToBytes()))

		switch mt := m.MessageType(); mt { //nolint:exhaustive
		case dhcpv4.MessageTypeDiscover:
			resp.UpdateOption(dhcpv4.OptMessageType(dhcpv4.MessageTypeOffer))
		case dhcpv4.MessageTypeRequest:
			resp.UpdateOption(dhcpv4.OptMessageType(dhcpv4.MessageTypeAck))
		default:
			log.Errorf("unhandled message type: %v", mt)

			return
		}

		log.Debug(resp.Summary())
		_, err = conn.WriteTo(resp.ToBytes(), peer)
		if err != nil {
			log.Printf("failure sending response: %s", err)
		}
	}
}

type DHCPLogger struct {
}

func (l DHCPLogger) PrintMessage(prefix string, message *dhcpv4.DHCPv4) {
	log.Infof("%s: %v", prefix, message)
}

func (l DHCPLogger) Printf(format string, v ...interface{}) {
	log.Infof(format, v...)
}

func (s *Server) startDHCP() error {
	return s.serverDHCP.Serve()
}

func getAvailableRange(netIp net.IPNet, netServer net.IP) (net.IP, net.IP) {
	mask := binary.BigEndian.Uint32(netIp.Mask)
	start := binary.BigEndian.Uint32(netServer.To4())

	first := start + 1
	last := ((start & mask) | (mask ^ 0xffffffff)) - 1

	firstIp := make(net.IP, 4)
	lastIp := make(net.IP, 4)

	binary.BigEndian.PutUint32(firstIp, first)
	binary.BigEndian.PutUint32(lastIp, last)

	return firstIp, lastIp
}


tftp.go
// Copyright 2016 Google Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"bytes"
	"errors"
	"fmt"
	"github.com/pin/tftp"
	"io"
	"io/ioutil"
	"net"
	"path/filepath"
	"strings"
)

var (
	ipxeFileName = "ipxe.efi"
)

type TFTPHook struct {
}

func (h *TFTPHook) OnSuccess(stats tftp.TransferStats) {
	log.Infof("Transferred %s to %s", stats.Filename, stats.RemoteAddr)
}

func (h *TFTPHook) OnFailure(stats tftp.TransferStats, err error) {
	log.Errorf("TFTPHook Failure transferring %s to %s: %s", stats.Filename, stats.RemoteAddr, err)
	if strings.Contains(err.Error(), "use of closed network connection") {
		//time.Sleep(time.Second)
	}
}

// readHandlerTFTP is called when client starts file download from server
func (s *Server) readHandlerTFTP(path string, rf io.ReaderFrom) error {
	_, classId, classInfo, err := extractInfo(path)
	if err != nil {
		return fmt.Errorf("unknown path %q", path)
	}

	bs, err := s.prepIpxeContent(classId, classInfo)
	if err != nil {
		return err
	}

	rf.(tftp.OutgoingTransfer).SetSize(int64(len(bs)))
	_, _ = rf.ReadFrom(bytes.NewBuffer(bs))

	return nil
}

// prepIpxeContent serves ipxe menu or the ipxe.efi file
func (s *Server) prepIpxeContent(classId, classInfo string) ([]byte, error) {

	if strings.Contains(classInfo, "iPXE") {
		var menuBuffer bytes.Buffer
		_ = ipxeMenuTemplate.Execute(&menuBuffer, s)
		return menuBuffer.Bytes(), nil
	}

	if classId == "PXEClient:Arch:00000:UNDI:002001" || classId == "PXEClient:Arch:00007:UNDI:003001" {
		data, err := ioutil.ReadFile(filepath.Join(s.ServerRoot, ipxeFileName))
		if err != nil {
			return nil, err
		}
		return data, nil
	}

	return nil, fmt.Errorf("Unknown class %s:%s", classId, classInfo)
}

func (s *Server) serveTFTP(l net.PacketConn) error {
	s.tFTPStarted = true
	if err := s.serverTFTP.Serve(l); err != nil {
		return fmt.Errorf("TFTP server shut down: %s", err)
	}
	return nil
}

func extractInfo(path string) (net.HardwareAddr, string, string, error) {
	pathElements := strings.Split(path, "/")
	if len(pathElements) != 3 {
		return nil, "", "", errors.New("not found")
	}

	mac, err := net.ParseMAC(pathElements[0])
	if err != nil {
		return nil, "", "", fmt.Errorf("invalid MAC address %q", pathElements[0])
	}

	classId := pathElements[1]
	classInfo := pathElements[2]

	return mac, classId, classInfo, nil
}

func (s *Server) logInfo(msg string) {
	log.Info(msg)
}
